### PR TITLE
build: rewrite tools/ scripts to deno

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes https://github.com/denoland/deno/issues/8223

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -93,14 +98,15 @@ jobs:
...
+      - name: lint.js
         if: matrix.kind == 'lint'
-        run: python ./tools/lint.py
+        run: deno run --unstable -A ./tools/lint.js
 
-      - name: test_format.py
+      - name: test_format.js
         if: matrix.kind == 'lint'
-        run: python ./tools/test_format.py
+        run: deno run --allow-run ./tools/test_format.js
```
Comment (bartlomieju) [inline .github/workflows/ci.yml:109]: Add `--check` to `tools/format.js`, no need to have separate scripts for that

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -133,7 +139,7 @@ jobs:
           DENOBOT_PAT: ${{ secrets.DENOBOT_PAT }}
         run: |
           git clone --depth 1 -b gh-pages https://${DENOBOT_PAT}@github.com/denoland/benchmark_data.git gh-pages
-          python ./tools/build_benchmark_jsons.py --release
+          deno run --unstable -A ./tools/build_benchmark_jsons.js --release
```
Comment (bartlomieju) [inline .github/workflows/ci.yml:142]: Check that this command succeeds before merging

File: tools/format.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+async function getSources(baseDir, patterns) {
+  const stagedOnly = Deno.args.includes("--staged");
+
+  if (stagedOnly) {
+    return await gitStaged(baseDir, patterns);
+  } else {
+    return await gitLsFiles(baseDir, patterns);
+  }
+}
```
Comment (bartlomieju) [inline tools/format.js:18]: Duplicated in `lint.js`, move to `util.js`

File: tools/format.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+async function dprint() {
+  const execPath = getPrebuiltToolPath("dprint");
+  console.log("dprint");
+  const p = Deno.run({
+    cmd: [execPath, "fmt"],
+  });
```
Comment (bartlomieju) [inline tools/format.js:25]: This is only command that doesn't pass files explicitly, probably an overlook and it should be changed

File: tools/util.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  const output = new TextDecoder().decode(await p.output());
+  p.close();
+
+  const files = output.split("\0").filter((line) => line.length > 0).map(
+    (filePath) => {
+      return Deno.realPathSync(join(baseDir, filePath));
+    },
+  );
+
+  return files;
```
Comment (bartlomieju) [inline tools/util.js:93]: Duplicated in `getLsFiles` - merge into a single function

**Review by james-and — APPROVED** [review]

**Review by james-and — APPROVED** [review]

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -73,6 +73,11 @@ jobs:
           rustup component add clippy
           rustup component add rustfmt
 
+      - name: Install Deno
+        uses: denolib/setup-deno@v2
+        with: 
+          deno-version: v1.5.1
```
Comment (bartlomieju) [inline .github/workflows/ci.yml:79]: Use `deno_install` script before landing

**Review by james-and — APPROVED** [review]

File: tools/format.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
```
Comment (bartlomieju) [inline tools/format.js:1]: Add shebang: `#!/usr/bin/env -S deno run -A`

File: tools/lint.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
```
Comment (bartlomieju) [inline tools/lint.js:1]: Ditto

File: .github/workflows/ci.yml (hunk 2 above)
Comment (bartlomieju) [reply .github/workflows/ci.yml:142]: Tried locally, working as expected.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -73,6 +73,11 @@ jobs:
           rustup component add clippy
           rustup component add rustfmt
 
+      - name: Install Deno
+        run: |-
+          curl -fsSL https://deno.land/x/install/install.sh | sh -s v1.5.1
+          echo "$HOME/.deno/bin" >> $GITHUB_PATH
+
```
Comment (bartlomieju) [inline .github/workflows/ci.yml:80]: Not sure how to make it work on Windows 😢

File: tools/format.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env -S deno run --unstable --allow-write --allow-read --allow-run
```
Comment (ry) [inline tools/format.js:1]: Unfortunately this doesn't work on linux

Comment (bartlomieju) [reply tools/format.js:1]: That's unfortunate :( it does work on `macOS`. I'm not super happy that incantations for linting and formatting will be so long but I think it's worth given we can remove all Python dependencies

**Review by ry — APPROVED** [review]
LGTM - nice work.

File: tools/format.js (hunk 10 above)
Comment (lucacasonato) [reply tools/format.js:1]: @ry It does with newer versions of /usr/bin/env. (ie ubuntu 19.04 and up).

File: tools/README.md [hunk 11]
```diff
@@ -2,24 +2,24 @@
...
+prerequisite to run this before code check in.
 
 To run formatting:
 
 ```sh
-./tools/format.py
+deno run -A --unstable ./tools/format.js
 ```
 
 ## lint.py
```
Comment (kt3k) [inline tools/README.md:16]: 👀

File: tools/format.js (hunk 10 above)
Comment (avindra) [reply tools/format.js:1]: `env` got -S (to split a single argument string into multiple arguments) in coreutils 8.30.

Some linux distros have 8.29 or lower, you can check [repology for coreutils](https://repology.org/project/coreutils/versions) for a fairly complete list of major linux package repos.

**Pull request merged**